File: Rust/src/utils/mat4.rs
use super::vec2::Vector2;

// Matrix4 { col1: Vector2, col2: Vector2 }
pub struct Matrix4 {
  pub col1: Vector2,
  pub col2: Vector2
}

impl Matrix4 {
  // Matrix4::new()
  pub fn new() -> Self {
    Self {
      col1: Vector2 { x: 0.0, y: 0.0 },
      col2: Vector2 { x: 0.0, y: 0.0 }
    }
  }

  // Matrix4 = Matrix4::from_angle(f32)
  pub fn from_angle(angle: f32) -> Self {
    let angle_cos = angle.cos();
    let angle_sin = angle.sin();

    let col1 = Vector2::new( angle_cos, angle_sin);
    let col2 = Vector2::new(-angle_sin, angle_cos);

    Self {
      col1: col1,
      col2: col2
    }
  }

  // Matrix4 = Matrix4::from_vectors(&Vector2, &Vector2)
  pub fn from_vectors(col1: &Vector2, col2: &Vector2) -> Self {
    Self {
      col1: Vector2 { x: col1.x, y: col1.y },
      col2: Vector2 { x: col2.x, y: col2.y }
    }
  }

  // Matrix4 = Matrix4.transpose()
  pub fn transpose(self) -> Self {
    Self {
      col1: Vector2::new(self.col1.x, self.col2.x),
      col2: Vector2::new(self.col1.y, self.col2.y)
    }
  }

  // Matrix4 = Matrix4.invert()
  pub fn invert(self) -> Self {
    let x1 = self.col1.x;
    let x2 = self.col2.x;
    let y1 = self.col1.y;
    let y2 = self.col2.y;

    let mut inv = Matrix4::new();
    let mut det = x1 * y2 - x2 * y1;

    debug_assert!(det != 0.0,
      "Matrix4::invert - 'det' is 0.0!"
    );

    det = 1.0 / det;

    inv.col1.x =  det * y2;
    inv.col2.x = -det * x2;
    inv.col1.y = -det * y1;
    inv.col2.y =  det * x1;

    return inv;
  }
}

File: Zig/src/utils/mat4.zig
const std = @import("std");
const Vec2 = @import("vec2.zig").Vec2;

pub const Mat4 = packed struct
{
    col1: Vec2 = Vec2 {},
    col2: Vec2 = Vec2 {},

    pub fn fromAngle(this: *Mat4, angle: f32) void
    {
        const cos = @cos(angle);
        const sin = @sin(angle);

        this.col1.x = cos;
        this.col1.y = sin;

        this.col2.x = -sin;
        this.col2.y =  cos;
    }

    pub fn fromVectors(this: *Mat4, col1: *const Vec2, col2: *const Vec2) void
    {
        this.col1.set(col1.x, col1.y);
        this.col2.set(col2.x, col2.y);
    }

    pub fn mul(this: *Mat4, multiplicand: *const Mat4) Mat4
    {
        var mat = Mat4 {};

        const col1 = this.mulVec2(&multiplicand.col1);
        const col2 = this.mulVec2(&multiplicand.col2);

        mat.fromVectors(&col1, &col2);
        return mat;
    }

    pub fn mulVec2(this: *Mat4, vec: *const Vec2) Vec2
    {
        return Vec2 {
            .x = this.col1.x * vec.x + this.col2.x * vec.y,
            .y = this.col1.y * vec.x + this.col2.y * vec.y
        };
    }

    pub fn add(this: *Mat4, addend: *const Mat4) Mat4
    {
        var mat = Mat4 {};

        const col1 = this.col1.add(&addend.col1);
        const col2 = this.col2.add(&addend.col2);

        mat.fromVectors(&col1, &col2);
        return mat;
    }

    pub fn transpose(this: *Mat4) Mat4
    {
        var mat = Mat4 {};

        const col1 = Vec2 { .x = this.col1.x, .y = this.col2.x };
        const col2 = Vec2 { .x = this.col1.y, .y = this.col2.y };

        mat.fromVectors(&col1, &col2);
        return mat;
    }

    pub fn invert(this: *Mat4) Mat4
    {
        const a = this.col1.x;
        const b = this.col2.x;
        const c = this.col1.y;
        const d = this.col2.y;

        var mat = Mat4 {};
        var det = a * d - b * c;
        std.debug.assert(det != 0.0);
        det = 1.0 / det;

        mat.col1.x =  det * d;
		mat.col1.y = -det * c;
        mat.col2.x = -det * b;
        mat.col2.y =  det * a;

        return mat;
    }

    // pub fn abs(this: *Mat4) void
    // {
    //     var col1 = this.col1.abs();
    //     var col2 = this.col2.abs();
    //     this.fromVectors(&col1, &col2);
    // }
};

test "basic functionality"
{
    var mat = Mat4 {};

    try std.testing.expect(
        mat.col1.x == 0 and
        mat.col1.y == 0 and
        mat.col2.x == 0 and
        mat.col2.y == 0
    );

    mat.fromAngle(0);

    try std.testing.expect(
        mat.col1.x == 1 and
        mat.col1.y == 0 and
        mat.col2.x == 0 and
        mat.col2.y == 1
    );

    var v1 = Vec2 { .x = 32, .y = -8 };
    var v2 = Vec2 { .x = -2, .y = 16 };

    mat.fromVectors(&v1, &v2);

    try std.testing.expect(
        mat.col1.x == 32 and
        mat.col1.y == -8 and
        mat.col2.x == -2 and
        mat.col2.y == 16
    );

    var mat4 = Mat4 {};
    mat4.fromAngle(1);
    mat4 = mat.mul(&mat4);

    try std.testing.expect(
        mat4.col1.x ==  15.606731 and
        mat4.col1.y ==   9.141117 and
        mat4.col2.x == -28.007675 and
        mat4.col2.y ==  15.376604
    );

    var vec = Vec2 { .x = 0.5, .y = 1.5 };
    vec = mat.mulVec2(&vec);
    try std.testing.expect(vec.x == 13 and vec.y == 20);

    mat4 = mat.add(&mat4);

    try std.testing.expect(
        mat4.col1.x ==  47.60673   and
        mat4.col1.y ==   1.1411171 and
        mat4.col2.x == -30.007675  and
        mat4.col2.y ==  31.376604
    );

    mat = mat.transpose();

    try std.testing.expect(
        mat.col1.x == 32 and
        mat.col1.y == -2 and
        mat.col2.x == -8 and
        mat.col2.y == 16
    );

    // mat.abs();

    // try std.testing.expect(
    //     mat.col1.x == 32 and
    //     mat.col1.y ==  2 and
    //     mat.col2.x ==  8 and
    //     mat.col2.y == 16
    // );

    mat.fromAngle(1);
    mat = mat.invert();

    try std.testing.expect(
        mat.col1.x ==  0.54030234 and
        mat.col1.y == -0.8414711  and
        mat.col2.x ==  0.8414711  and
        mat.col2.y ==  0.54030234
    );
}
